ADBDEV-4059-1 Make InitPlans non-rescannable during estate initialization

bimboterminator1 opened the pull request:
The executor enabled the EXEC_FLAG_REWIND flag for all types of SubPlans: either
for InitPlans or correlated/uncorrelated SubPlans. This flag represents that the
rescan is expected and is used during initialization of executor state. However,
if a query had InitPlans, which contained non-rescannable nodes (like Split
Update in the tests), the executor failed with assertion error (for example,
when calling ExecInitSplitUpdate when initializing executor state inside the
ExecInitNode for the InitPlan).

Because InitPlans are essentially executed only once, there is no need to expect
a rescan of the InitPlan. Therefore, in order to support non-rescannable
operations inside the InitPlans, this patch disables EXEC_FLAG_REWIND flag for
InitPlans.

This patch partially returns vanilla postgres logic, which used
plannedstmt->rewindPlanIDs bitmapset for making a decision whether current
SubPlan should be executed with EXEC_REWIND flag. This bitmapset used to be
filled with the ids of such SubPlans, that could optimize the rescan
operation if the EXEC_REWIND is set, like parameterless subplans. Other
types of SubPlans were considered rescannable by default and there were
no need to set the EXEC_REWIND flag for them. However, GPDB interprets the
EXEC_REWIND flag as an indicator that the node is likely to be rescanned, and
also used this flag to delay the eager free. Therefore, this patch proposes
to fill plannedstmt->rewindPlanIDs set with all the subplans ids, except
InitPlans, and to set EXEC_REWIND flag only for those subplans that are
in the rewindPlanIDs bitmapset.

As for legacy optimizer, the if-clause influencing the filling of the
bitmapset is changed inside the build_subplan function in order to filter out
any InitPlans.

As for ORCA optimizer, rewindPlanIDs was not previously used, and this patch
adds a bunch of logic to fill this bitmapset with subplan ids. This
patch extends existing SetInitPlanVariables function and renames it to
SetSubPlanVariables. This function has originally been setting
nInitPlans and nParamExec in PlannedStmt, and also has been setting
qDispSliceId for each InitPlan, that is found during plan tree
traversal. This patch extends this behaviour and additionally fills the
rewindPlanIDs bitmapset for each SubPlan found, execept InitPlans.

At executor side, the condition checking whether the SubPlan is
in the planned_stmt->rewindPlanIDs is added to the InitPlan function.
From that point, SubPlans will be initialized with EXEC_REWIND flag
only if they are not InitPlans.

bimboterminator1 committed:
Make InitPlans non-rescannable

The executor enabled the EXEC_FLAG_REWIND flag for all types of SubPlans: either
for InitPlans or correlated/uncorrelated SubPlans. This flag represents that the
rescan is expected and is used during initialization of executor state. However,
if a query had InitPlans, which contained non-rescannable nodes (like Split
Update in the tests), the executor failed with assertion error (for example,
when calling ExecInitSplitUpdate when initializing executor state inside the
ExecInitNode for the InitPlan).

Because InitPlans are essentially executed only once, there is no need to expect
a rescan of the InitPlan. Therefore, in order to support non-rescannable
operations inside the InitPlans, this patch disables EXEC_FLAG_REWIND flag for
InitPlans. Before the ExecInitNode is called for a SubPlan, the if-clause, that
checks whether this SubPlan is an InitPlan, is added. This condition checks the
occurence of subplan_idx inside the pre collected set of InitPlan plan_id. This
set is formed by executing additional walker getInitPlans, that collects all
InitPlan ids from the plan tree.

Stolb27 reviewed (commented):
@bimboterminator1, can you explain how these non-rescannable nodes handled in sub plans? Does gpdb produce a valid plan in this case?

bimboterminator1 commented:
> @bimboterminator1, can you explain how these non-rescannable nodes handled in sub plans? Does gpdb produce a valid plan in this case?

For correlated SubPlans the same issue occurs (for SplitUpdate):
```
explain (costs off)
with cte as
(update t1 set i = 0
 returning i)
select i from t2
where t2.i in (select i from cte where t2.i = cte.i);
FATAL:  Unexpected internal error (nodeSplitUpdate.c:176)
DETAIL:  FailedAssertion("!(!(eflags & (0x0004 | 0x0008 | 0x0002)))", File: "nodeSplitUpdate.c", Line: 176)
```
For update/delete with MOVEMENT_EXPLICIT another assert fails for correlated Subplan
```
explain (costs off)
with cte as
(delete from t1 returning i)
select i from t2
where t2.i in (select i from cte where t2.i = cte.i);
FATAL:  Unexpected internal error (cdbllize.c:1228)
DETAIL:  FailedAssertion("!(flow->req_move == MOVEMENT_NONE)", File: "cdbllize.c", Line: 1228)
```
For uncorrelated subplans everything is ok, because Modify Table nodes are materialized.

bimboterminator1 committed:
Add tests for ppc64le

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/52691

Stolb27 commented on src/backend/executor/execMain.c:
```diff
@@ -2264,11 +2267,13 @@ InitPlan(QueryDesc *queryDesc, int eflags)
 			sp_eflags = eflags
 				& (EXEC_FLAG_EXPLAIN_ONLY | EXEC_FLAG_WITH_NO_DATA);
-			sp_eflags |= EXEC_FLAG_REWIND;
+			if (!bms_is_member(subplan_idx, initplans))
+				sp_eflags |= EXEC_FLAG_REWIND;
```
Did you comapre your implementation with original implemenattion of Postgres 9.4?

```c
		 * A subplan will never need to do BACKWARD scan nor MARK/RESTORE. If
		 * it is a parameterless subplan (not initplan), we suggest that it be
		 * prepared to handle REWIND efficiently; otherwise there is no need.
		 */
		sp_eflags = eflags
			& (EXEC_FLAG_EXPLAIN_ONLY | EXEC_FLAG_WITH_NO_DATA);
		if (bms_is_member(i, plannedstmt->rewindPlanIDs))
			sp_eflags |= EXEC_FLAG_REWIND;
```

Moreover, this variable exists for gpdb and seems to be in use!

bimboterminator1 replied:
Well, first of all, we should look at gpdb [comment](https://github.com/arenadata/gpdb/blob/fcfac650cc8349ece9662dfeea6a34071f8d0db7/src/backend/executor/execMain.c#L2267):
```
 GPDB: We always set the REWIND flag, to delay eagerfree.
```
Therefore, EXEC_REWING should be set for any SubPlan.

Secondly, let's look at the only [usage](https://github.com/arenadata/gpdb/blob/fcfac650cc8349ece9662dfeea6a34071f8d0db7/src/backend/optimizer/plan/subselect.c#L990-L999) of `rewindPlanIDs`:
```
if (splan->parParam == NIL && !splan->is_initplan && !splan->useHashTable)
		root->glob->rewindPlanIDs = bms_add_member(root->glob->rewindPlanIDs,
				                           splan->plan_id);
```
In other words, some correlated SubPlans won't be added to rewindPlanIDs as well. And I guess EXEC_REWIND anyway is expected for them.

RekGRpth replied:
> if (splan->parParam == NIL && !splan->is_initplan && !splan->useHashTable)
> 		root->glob->rewindPlanIDs = bms_add_member(root->glob->rewindPlanIDs,
> 				                           splan->plan_id);

rewindPlanIDs list does not contain init plans, does it?

RekGRpth replied:
or you can add `Bitmapset  *initPlanIDs` member to `PlannerGlobal` struct and fill it by condition
```c
if (splan->is_initplan)
    root->glob->initPlanIDs = bms_add_member(root->glob->initPlanIDs, splan->plan_id);
```
and then check in it
```c
if (!bms_is_member(subplan_idx, plannedstmt->initPlanIDs))
    sp_eflags |= EXEC_FLAG_REWIND;
```

bimboterminator1 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-4059-1

bimboterminator1 commented on src/backend/executor/execMain.c:
```diff
@@ -2248,6 +2249,8 @@ InitPlan(QueryDesc *queryDesc, int eflags)
 		locallyExecutableSubplans = getLocallyExecutableSubplans(plannedstmt, start_plan_node);
 	}
 
+	initplans = getInitPlans(plannedstmt, start_plan_node);
```
Suggest such an optimization:
```suggestion
if (list_length(plannedstmt->subplans) > 0)
	initplans = getInitPlans(plannedstmt, start_plan_node);
```

bimboterminator1 committed:
Revert executor-side solution

This reverts commit c92b019c6becf1e80641543361468a3ded99b982.

bimboterminator1 committed:
Make InitPlans non-rescannable

The executor enabled the EXEC_FLAG_REWIND flag for all types of SubPlans: either
for InitPlans or correlated/uncorrelated SubPlans. This flag represents that the
rescan is expected and is used during initialization of executor state. However,
if a query had InitPlans, which contained non-rescannable nodes (like Split
Update in the tests), the executor failed with assertion error (for example,
when calling ExecInitSplitUpdate when initializing executor state inside the
ExecInitNode for the InitPlan).

Because InitPlans are essentially executed only once, there is no need to expect
a rescan of the InitPlan. Therefore, in order to support non-rescannable
operations inside the InitPlans, this patch disables EXEC_FLAG_REWIND flag for
InitPlans.

This patch partially returns vanilla postgres logic, which used
plannedstmt->rewindPlanIDs bitmapset for making a decision whether current
SubPlan should be executed with EXEC_REWIND flag. This bitmapset used to be
filled with the ids of such SubPlans, that could optimize the rescan
operation if the EXEC_REWIND is set, like parameterless subplans. Other
types of SubPlans were considered rescannable by default and there were
no need to set the EXEC_REWIND flag for them. However, GPDB interprets the
EXEC_REWIND flag as an indicator that the node is likely to be rescanned, and
also used this flag to delay the eager free. Therefore, this patch proposes
to fill plannedstmt->rewindPlanIDs set with all the subplans ids, except
InitPlans, and to set EXEC_REWIND flag only for those subplans that are
in the rewindPlanIDs bitmapset.

As for legacy optimizer, the if-clause influencing the filling of the
bitmapset is changed inside the build_subplan function in order to filter out
any InitPlans.

As for ORCA optimizer, rewindPlanIDs was not previously used, and this patch
adds a bunch of logic to fill this bitmapset with subplan ids. Firstly, the
CContextDXLToPlStmt class is extended with an extra member m_subplan_exprs_list
that consists of expressions ot SubPlan type. This list is filled via
AddSubPlanExpr member function, which is called from the
TranslateSubplanFromChildPlan function. Also the getter for the
m_subplan_exprs_list is added. Next, the CTranslatorDXLToPlStmt is
extended with SetRewindPlanIds function, that performs the filtration
of the m_subplan_exprs_list and fills the planned_stmt->rewindPlanIDs
with subplan ids, that correspond to any SubPlan, except InitPlans.
The SetRewindPlanIds function is called at one of the final planning
stages, where DXL plan tree in converted to the PlannedStmt. Thus,
proposed changes allows to correctly fill the planned_stmt->rewindPlanIDs
bitmapset.

At executor side, the condition checking whether the SubPlan is
in the planned_stmt->rewindPlanIDs is added to the function InitPlan.
From that point, SubPlans will be initialized with EXEC_REWIND flag
only if they are not InitPlans.

bimboterminator1 committed:
Make InitPlans non-rescannable

The executor enabled the EXEC_FLAG_REWIND flag for all types of SubPlans: either
for InitPlans or correlated/uncorrelated SubPlans. This flag represents that the
rescan is expected and is used during initialization of executor state. However,
if a query had InitPlans, which contained non-rescannable nodes (like Split
Update in the tests), the executor failed with assertion error (for example,
when calling ExecInitSplitUpdate when initializing executor state inside the
ExecInitNode for the InitPlan).

Because InitPlans are essentially executed only once, there is no need to expect
a rescan of the InitPlan. Therefore, in order to support non-rescannable
operations inside the InitPlans, this patch disables EXEC_FLAG_REWIND flag for
InitPlans.

This patch partially returns vanilla postgres logic, which used
plannedstmt->rewindPlanIDs bitmapset for making a decision whether current
SubPlan should be executed with EXEC_REWIND flag. This bitmapset used to be
filled with the ids of such SubPlans, that could optimize the rescan
operation if the EXEC_REWIND is set, like parameterless subplans. Other
types of SubPlans were considered rescannable by default and there were
no need to set the EXEC_REWIND flag for them. However, GPDB interprets the
EXEC_REWIND flag as an indicator that the node is likely to be rescanned, and
also used this flag to delay the eager free. Therefore, this patch proposes
to fill plannedstmt->rewindPlanIDs set with all the subplans ids, except
InitPlans, and to set EXEC_REWIND flag only for those subplans that are
in the rewindPlanIDs bitmapset.

As for legacy optimizer, the if-clause influencing the filling of the
bitmapset is changed inside the build_subplan function in order to filter out
any InitPlans.

As for ORCA optimizer, rewindPlanIDs was not previously used, and this patch
adds a bunch of logic to fill this bitmapset with subplan ids. Firstly, the
CContextDXLToPlStmt class is extended with an extra member m_subplan_exprs_list
that consists of expressions ot SubPlan type. This list is filled via
AddSubPlanExpr member function, which is called from the
TranslateSubplanFromChildPlan function. Also the getter for the
m_subplan_exprs_list is added. Next, the CTranslatorDXLToPlStmt is
extended with SetRewindPlanIds function, that performs the filtration
of the m_subplan_exprs_list and fills the planned_stmt->rewindPlanIDs
with subplan ids, that correspond to any SubPlan, except InitPlans.
The SetRewindPlanIds function is called at one of the final planning
stages, where DXL plan tree in converted to the PlannedStmt. Thus,
proposed changes allows to correctly fill the planned_stmt->rewindPlanIDs
bitmapset.

At executor side, the condition checking whether the SubPlan is
in the planned_stmt->rewindPlanIDs is added to the function InitPlan.
From that point, SubPlans will be initialized with EXEC_REWIND flag
only if they are not InitPlans.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/54365

RekGRpth commented:
You changed the ORCA code. Could you add tests (unit or regress) to check these changes?

bimboterminator1 commented:
> You changed the ORCA code. Could you add tests (unit or regress) to check these changes?

It seems that ORCA can't create InitPlans at all. I didn't manage to find any example of ORCA building InitPlan.
```
explain (costs off)
select * from t2 where 0 =  (select t_strewn.j from t_strewn);
                       QUERY PLAN                       
--------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)
   ->  Result
         One-Time Filter: (0 = $0)
         InitPlan 1 (returns $0)  (slice3)
           ->  Gather Motion 3:1  (slice1; segments: 3)
                 ->  Seq Scan on t_strewn
         ->  Seq Scan on t2
 Optimizer: Postgres query optimizer
(8 rows)

postgres=# set optimizer = 1;
SET
postgres=# explain (costs off)
select * from t2 where 0 =  (select t_strewn.j from t_strewn);
                              QUERY PLAN                               
-----------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)
   ->  Seq Scan on t2
         Filter: (SubPlan 1)
         SubPlan 1  (slice2; segments: 3)
           ->  Result
                 ->  Materialize
                       ->  Broadcast Motion 3:3  (slice1; segments: 3)
                             ->  Seq Scan on t_strewn
                                   Filter: (0 = j)
 Optimizer: Pivotal Optimizer (GPORCA)
(10 rows)
```
And the variable `is_initplan` is always `false` when SubPlan is [created](https://github.com/arenadata/gpdb/blob/fcfac650cc8349ece9662dfeea6a34071f8d0db7/src/backend/gpopt/translate/CTranslatorDXLToScalar.cpp#L1036). I  haven't found any place where `is_initplan` is set to `true`  in ORCA.

RekGRpth commented on src/backend/executor/execMain.c:
```diff
@@ -2264,11 +2264,13 @@ InitPlan(QueryDesc *queryDesc, int eflags)
 			sp_eflags = eflags
 				& (EXEC_FLAG_EXPLAIN_ONLY | EXEC_FLAG_WITH_NO_DATA);
-			sp_eflags |= EXEC_FLAG_REWIND;
+			if (bms_is_member(subplan_idx + 1, plannedstmt->rewindPlanIDs))
```
If ORCA does not generate init plans, how about such
```suggestion
			if (plannedstmt->planGen == PLANGEN_OPTIMIZER || bms_is_member(subplan_idx + 1, plannedstmt->rewindPlanIDs))
```
instead all code changes in ORCA?

bimboterminator1 replied:
Done

Stolb27 replied:
Can you explain how absence of init-plan generation in ORCA is connected with adding one more kludge with planner-dependent logic in the executor? What problems with initial solution, that modifying ORCA?

bimboterminator1 replied:
>What problems with initial solution, that modifying ORCA?

Do we need to multiply entities beyond necessity?

Stolb27 replied:
If it makes code more straightforward, I think we can. We have `rewindPlanIDs` in the Planned Statement structure. Do we really want to care about the kind of planner that produce this plan to understand that we can rely on the value stored in this field? I could understand you if it requires many efforts. The less the executor knows about the specific of the exact planner, the better.

bimboterminator1 replied:
Fair point. I'll restore ORCA changes

bimboterminator1 committed:
Revert ORCA changes

This reverts commit c83ff195a654708a0e9b14b9b2f36c9965e99f72.

bimboterminator1 committed:
Make InitPlans non-rescannable

The executor enabled the EXEC_FLAG_REWIND flag for all types of SubPlans: either
for InitPlans or correlated/uncorrelated SubPlans. This flag represents that the
rescan is expected and is used during initialization of executor state. However,
if a query had InitPlans, which contained non-rescannable nodes (like Split
Update in the tests), the executor failed with assertion error (for example,
when calling ExecInitSplitUpdate when initializing executor state inside the
ExecInitNode for the InitPlan).

Because InitPlans are essentially executed only once, there is no need to expect
a rescan of the InitPlan. Therefore, in order to support non-rescannable
operations inside the InitPlans, this patch disables EXEC_FLAG_REWIND flag for
InitPlans.

This patch turns back to vanilla postgres field of PlannedStmt structure,
plannedstmt->rewindPlanIDs bitmapset, that was used for making a decision
whether current SubPlan should be executed with EXEC_REWIND flag. This bitmapset
used to be filled with the ids of such SubPlans, that could optimize the rescan
operation if the EXEC_REWIND is set, like parameterless subplans. Other
types of SubPlans were considered rescannable by default and there were
no need to set the EXEC_REWIND flag for them. However, GPDB interprets the
EXEC_REWIND flag as an indicator that the node is likely to be rescanned, and
also uses this flag to delay the eager free. Therefore, this patch proposes
to fill plannedstmt->rewindPlanIDs set with all the subplans ids, except
InitPlans, and to set EXEC_REWIND flag only for those subplans that are
in the rewindPlanIDs bitmapset.

For legacy optimizer, the if-clause influencing the filling of the
bitmapset is changed inside the build_subplan function in order to filter out
any InitPlans.
As for ORCA optimizer, rewindPlanIDs was not previously used, and, moreover,
ORCA is unable to produce InitPlans. Therefore, all SubPlans for plans
generated by the optimizer must be initialized with EXEC_REWIND flag.

At executor side, the condition checking whether the SubPlan is
in the planned_stmt->rewindPlanIDs is added to the InitPlan function.
For ORCA generated plans the filtration is not held, and EXEC_REWIND
is always set as it was before.
From that point, SubPlans will be initialized with EXEC_REWIND flag
only if they are not InitPlans.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/54518

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/54549

Stolb27 commented on src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:
```diff
@@ -463,6 +465,30 @@ CTranslatorDXLToPlStmt::SetParamIds(Plan *plan)
+			bitmapset = gpdb::BmsAddMember(bitmapset, subplan->plan_id);
+	}
+
+	planned_stmt->rewindPlanIDs = bitmapset;
```
Did you consider `CTranslatorDXLToPlStmt::SetInitPlanVariables` as a place for this logic? This function traverse through plan and collect all subplans (including init plans too).

bimboterminator1 replied:
Yes, I did. This function suits best for our purposes. However, how legal is it to change the function's 1) naming, 2) specification and comments.

We can make the ` SetRewindPlanIds` function traverse the whole tree to find InitPlans as  in `SetInitPlanVariables`. But I think it's more expensive, than maintain the state list ` m_subplan_exprs_list`.

bimboterminator1 replied:
Thus, what is better:  change `SetInitPlanVariables` and fill the rewindPlansIDs during tree traversal, or add  `SetRewindPlanIds` function? And if we add a separate function is it better to maintain another list of SubPlan's or implement another tree traversal?

Stolb27 replied:
In my opinion, the presence of both approaches together looks redundant. If you want to implement a separate list of subplans, you need to rework `SetInitPlanVariables` too. But, it's more simple to change function name (e.g. to `SetSubPlanVariables`) and description and collect all non-init-plans to bitmap here.

bimboterminator1 replied:
Done

bimboterminator1 committed:
Make InitPlans non-rescannable

The executor enabled the EXEC_FLAG_REWIND flag for all types of SubPlans: either
for InitPlans or correlated/uncorrelated SubPlans. This flag represents that the
rescan is expected and is used during initialization of executor state. However,
if a query had InitPlans, which contained non-rescannable nodes (like Split
Update in the tests), the executor failed with assertion error (for example,
when calling ExecInitSplitUpdate when initializing executor state inside the
ExecInitNode for the InitPlan).

Because InitPlans are essentially executed only once, there is no need to expect
a rescan of the InitPlan. Therefore, in order to support non-rescannable
operations inside the InitPlans, this patch disables EXEC_FLAG_REWIND flag for
InitPlans.

This patch partially returns vanilla postgres logic, which used
plannedstmt->rewindPlanIDs bitmapset for making a decision whether current
SubPlan should be executed with EXEC_REWIND flag. This bitmapset used to be
filled with the ids of such SubPlans, that could optimize the rescan
operation if the EXEC_REWIND is set, like parameterless subplans. Other
types of SubPlans were considered rescannable by default and there were
no need to set the EXEC_REWIND flag for them. However, GPDB interprets the
EXEC_REWIND flag as an indicator that the node is likely to be rescanned, and
also used this flag to delay the eager free. Therefore, this patch proposes
to fill plannedstmt->rewindPlanIDs set with all the subplans ids, except
InitPlans, and to set EXEC_REWIND flag only for those subplans that are
in the rewindPlanIDs bitmapset.

As for legacy optimizer, the if-clause influencing the filling of the
bitmapset is changed inside the build_subplan function in order to filter out
any InitPlans.

As for ORCA optimizer, rewindPlanIDs was not previously used, and this patch
adds a bunch of logic to fill this bitmapset with subplan ids. This
patch extends existing SetInitPlanVariables function and renames it to
SetSubPlanVariables. This function has originally been setting
nInitPlans and nParamExec in PlannedStmt, and also has been setting
qDispSliceId for each InitPlan, that is found during plan tree
traversal. This patch extends this behaviour and additionally fills the
rewindPlanIDs bitmapset for each SubPlan found, execept InitPlans.

At executor side, the condition checking whether the SubPlan is
in the planned_stmt->rewindPlanIDs is added to the InitPlan function.
From that point, SubPlans will be initialized with EXEC_REWIND flag
only if they are not InitPlans.

bimboterminator1 committed:
Formating

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/54638

Stolb27 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-4059-1

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/54666
